File: Python/Challenge7.py
#
#   Reto #7
#   CONTANDO PALABRAS
#   Fecha publicación enunciado: 14/02/22
#   Fecha publicación resolución: 21/02/22
#   Dificultad: MEDIA
#  
#   Enunciado: Crea un programa que cuente cuantas veces se repite cada palabra y que muestre el recuento final de todas
#   ellas.
#   - Los signos de puntuación no forman parte de la palabra.
#   - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
#   - No se pueden utilizar funciones propias del lenguaje que lo resuelvan automáticamente.
#  
#   Información adicional:
#   - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar
#     ayuda la comunidad.
#   - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
#   - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
#   - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
#  

def has_punctuation(phrase: str) -> list:
    """
    Función que realiza la extracción de símbolos de puntuación de una frase, y separa en palabras.
    """

    # Símbolos de puntuación que serán eliminados
    punctuation_chars = [".", ",", ":", ";", "!", "¡", "?", "¿", "(", ")", "{", "}", "[", "]", "-"]
    new_word = ""

    for char in phrase:
        # Si el caracter en la frase se encuentra dentro del listado de puntuación, lo elimina
        if char in punctuation_chars:
            char = ""

        # Concatena el caracter en un String.
        new_word += char

    # El bucle For anterior, se pudo haber sustituido por:
    #
    # for _ in punctuation_chars:
    #   phrase = phrase.replace(_, "")
    # Return phrase.lower().split(" ")
    #

    return new_word.lower().split(" ")


def count_words(phrase: str):
    """
    Cuenta las palabras dentro de una frase, **Toma** en cuenta la acentuación, y las imprime en la consola.

    Ejemplo: 'Él' != 'El', por ende, son dos palabras distintas.
    """
    words_list = has_punctuation(phrase)
    count = {}

    for word in words_list:
        # Si la palabra no está en el diccionario de palabras, la añade con una repetición de 1.
        if word not in count:
            count[word] = 1
        # En caso de que exista en el diccionario, le añade una repetición.
        else:
            count[word] += 1

    # Imprime los resultados de conteo de palabras
    print("El conteo de palabras es: ")
    for register in count:
        print(f"{register}: {count[register]}")


# Run Test
count_words("Y así como suele decirse el gato al rato, el rato a la cuerda, la cuerda al palo; daba el arriero a Sancho, Sancho a la moza, la moza a él, el ventero a la moza")

# Result Test
# El conteo de palabras es:
# y: 1
# así: 1
# como: 1
# suele: 1
# decirse: 1
# el: 4
# gato: 1
# al: 2
# rato: 2
# a: 5
# la: 5
# cuerda: 2
# palo: 1
# daba: 1
# arriero: 1
# sancho: 2
# moza: 3
# él: 1
# ventero: 1

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge7.rb
# frozen_string_literal: true

require 'set'
#
#  Reto #7
#  CONTANDO PALABRAS
#  Fecha publicación enunciado: 14/02/22
#  Fecha publicación resolución: 21/02/22
#  Dificultad: MEDIA
#
#  Enunciado: Crea un programa que cuente cuantas veces se repite cada
#  palabra y que muestre el recuento final de todas ellas.
#  - Los signos de puntuación no forman parte de la palabra.
#  - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
#  - No se pueden utilizar funciones propias del lenguaje que lo resuelvan
#    automáticamente.
#
#  Información adicional:
#  - Usa el canal de nuestro discord (https://mouredev.com/discord)
#    "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la acomunidad.
#  - Puedes hacer un Fork del repo y una Pull Request al repo original para
#    que veamos tu solución aportada.
#  - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de
#    su publicación.
#  - Subiré una posible solución al ejercicio el lunes siguiente al de su
#    publicación.
#

# Solución:
# Requerimientos:
# Ruby 2.7.3
#
# Ejecución:
# ruby Challenge7_test.rb

# class to call method
class Main
  def initialize(string)
    @string = string
  end

  def execute
    count_string = StringTool.new(@string)
    count_string.counting_repetitions
  end
end

# class that contains the method
class StringTool
  def initialize(string)
    @string = string
  end

  def counting_repetitions
    new_string = downcase_string(@string)
    clean_string = only_words(new_string)
    # clean_string.split(' ').group_by(&:itself).map { |k, v| [k, v.count] }.to_h
    # clean_string.split(' ').group_by(&:itself).transform_values(&:count)

    # using string procedures
    collection = Set.new(clean_string.split(' ')).to_a
    result = {}
    collection.each do |element|
      result[element] = clean_string.scan(element).size
    end
    result
  end

  private

  def downcase_string(string)
    if string.chars.any?(/[ÁÉÍÓÚ]/)
      string.downcase
    else
      # procedure can't downcase characters ÁÉÍÓÚ
      string.chars.map { |c| c.match?(/[A-Z]/) ? (c.ord + 32).chr : c }.join('')
    end
  end

  def only_words(string)
    string.gsub(/[^a-z0-9áéíóúÁÉÍÓÚ\s]/, '')
  end
end
